Add checks for forgetting Module() wraps or double Module() wrapping

jackkoenig opened the pull request:
Implemented by changing currentModule to moduleStack
Fixes #192 

Deleted DeqIOSpec because it wasn't really doing anything and it was missing a Module() wrap anyway. When I added the Module() wrap the test just times out.

jackkoenig commented:
Note that this might break other repos' tests because even in this repo people sometimes forget to Module wrap or double Module wrap and things just kinda happened to work out.

aswaterman commented on chiselFrontend/src/main/scala/chisel3/core/Module.scala:
```diff
@@ -25,16 +25,22 @@ object Module {
+      throwException("Error: Top of Module stack is not the Module expected. " +
+                     "You probably forgot a Module() wrapper or double Module() wrapped.")
+
+    }
```
I suggest splitting these cases apart and giving the more precise error message in either case.

aswaterman replied:
Ah, I see.

sdtwigg commented:
I think you can do this without a Stack by just having a global boolean variable.

Module() asserts variable is false (and then sets it to true) and Module constructor asserts variable is true (and then sets it to false). If I recall, this is what was done in Chisel 2 with modStackPushed.

sdtwigg commented:
Amendment: Module() should check twice. Before setting the variable and after constructing the Module. The latter check catches trying to re-Module an already instantiated Module (which is like double wrapping but different due to the fact Module() is call-by-name):
```scala
val myMod = Module(new Blah)
val myMod2 = Module(myMod)
```

aswaterman commented:
Agreed, but the intent is arguably clearer this way, and it's conceivable that some future implementation feature would want the explicit stack anyway.  (And, of course, the performance implications are irrelevant.)

jackkoenig commented:
Alright done. Switch implementation to boolean because it can also detect rewrapping with Module. Also switched back to currentModule because the stack was no longer being used (can always be reimplemented if it's useful). Also added useful function on SourceLine for printing error messages.
